docs/design/password-key-derivation.md

```diff
@@ -0,0 +1,66 @@
+# Title: Password Key Derivation
+
+## Abstract
+
+This design doc describes the way cryptographic encryption keys will be derived from user provided passwords in a way that resists offline attacks and low entropy passwords.
+
+## Background
+
+A unique encryption key exists for every path inside of a bucket, and they are designed in a way that allows deriving a key from a parent and some path component. For example, if you had the key for `sj://bucket/path1/path2` then you could derive the key for `sj://bucket/path1/path2/path3`.
+
+That leaves open the question for how a root key is derived. Some requirements on that root key derivation include
+
+1. It's based on the user's choice (for example, a password).
+2. If the user inputs the same password on different machines, the same root key should be derived.
+
+These requirements allow users to be in full control of their encryption, and don't require users safely transporting high entropy (hard to remember) secrets to bootstrap new uplinks.
+
+This design accomodates more requirements that allow for additional features:
+
+3. A root key can be derived for any path in a bucket.
+4. Buckets root keys should not depend on the name of the bucket.
+5. A table of keys for low entropy passwords should not be possible. In other words, an attacker with knowledge of the algorithm should not be able to use a dictionary of common passwords and pre-compute what keys to check in the event of a data breach.
+6. There is a default that allows deriving keys for buckets when no bucket specific secret is provided.
+
+The third requirement allows having multiple encrypted domains to exist within a single bucket, allowing delegation of encryption. The fourth requirement allows renaming buckets without having to re-encrypt all of the paths inside of a bucket. The fifth requirement enhances security if a satellite breach happens. This is usually accomplished with [salting](https://en.wikipedia.org/wiki/Salt_(cryptography)). The sixth requirement allows providing a finite amount of information to a third party that allows them access to the keys for any bucket created in the future.
+
+## Design
+
+A default password can be chosen. There is no extra entropy or information available, so it will be run through [Argon2](https://en.wikipedia.org/wiki/Argon2) with a very high work load and a fixed salt. The output of that will be stored in a secure location. The higher the workload, the better, as it increases the cost for an offline attack, and only the output is used in subsequent steps, so it only needs to happen once.
+
+Satellites will add a new field to bucket metadata called `entropy` that just contains 32 random, stable bytes.
```

Comment:
What about using the project UUID as the salt, to avoid having to ask the server for it? I don't recall whether we were planning to include that in our sharing bundles.

Reply:
It's not included in any sharing bundles or api key or otherwise. It would require doing an extra network request over what we already do.

Reply:
It would not require any network access if we added the project UUID to sharing bundles or api keys. Or just a prefix of the project UUID.

Reply:
Indeed, at the cost of making the share larger. I agree this will work if we can't piggyback, but I believe we can.

Reply:
That does make the API key longer

Reply:
Given that we need to ask the server anyway, I'm fine with per-bucket entropy. As a final question, should we not add that as a field to the bucket-level pointer object, since we need to fetch that anyway for the bucket metadata? Or are bucket-level pointers going away to be replaced by an expanded nodes table? I remember discussing that at one point.

Reply:
Another unrelated question: would it not be better to rename this `entropy` field to `bucketID` or `bucketUUID`? Because this field is essentially becoming a persistent ID to take the place of the bucket name, which will no longer be permanent. Besides being more self-explanatory in purpose, it also makes it more obvious that the bytes contained by the field are entirely public and don't need to be guarded or anything.

OK, yes, this does not actually meet the criterion of being a total showstopper for me, but I do think it would aid future spelunking expeditions.

Reply:
The proposal has changed since this last comment was made to be an explicit call. I'll happily take suggestions on what to name it.

Reply:
I'm fine with GetSaltEntropy.